File: 918-maximum-sum-circular-subarray/solution.py
class Solution:
  def maxSubarraySumCircular(self, nums: list[int]) -> int:
    total = local_max = global_max = local_min = global_min = nums[0]
    for i in range(1, len(nums)):
      num = nums[i]
      total += num
      local_max = max(0, local_max) + num
      global_max = max(global_max, local_max)
      local_min = min(0, local_min) + num
      global_min = min(global_min, local_min)

    if total == global_min:
      return global_max

    return max(global_max, total - global_min)

File: 918-maximum-sum-circular-subarray/solution.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
  int maxSubarraySumCircular(std::vector<int> &nums)
  {
    int total, local_max, global_max, local_min, global_min;
    total = local_max = global_max = local_min = global_min = nums[0];
    for (size_t i = 1; i < nums.size(); ++i)
    {
      const int num = nums[i];
      total += num;
      local_max = std::max(0, local_max) + num;
      global_max = std::max(global_max, local_max);
      local_min = std::min(0, local_min) + num;
      global_min = std::min(global_min, local_min);
    }

    if (total == global_min)
    {
      return global_max;
    }

    return std::max(global_max, total - global_min);
  }
};